src/lib/keccak.unit-test.ts

```diff
@@ -1,106 +1,136 @@
+import { array, equivalentAsync, fieldWithRng } from './testing/equivalent.js';
+import { constraintSystem, contains } from './testing/constraint-system.js';
 
-const KECCAK_TEST_STATE = [
-  [0, 0, 0, 0, 0],
-  [0, 0, 1, 0, 0],
-  [1, 0, 0, 0, 0],
-  [0, 0, 0, 1, 0],
-  [0, 1, 0, 0, 0],
-].map((row) => row.map((elem) => Field.from(elem)));
+const PREIMAGE_LENGTH = 75;
+const RUNS = 1;
 
-let KeccakBlockTransformation = ZkProgram({
-  name: 'KeccakBlockTransformation',
-  publicInput: Provable.Array(Provable.Array(Field, KECCAK_DIM), KECCAK_DIM),
-  publicOutput: Provable.Array(Provable.Array(Field, KECCAK_DIM), KECCAK_DIM),
+let uint = (length: number) => fieldWithRng(Random.biguint(length));
 
+let Keccak256 = ZkProgram({
+  name: 'keccak256',
+  publicInput: Provable.Array(Field, PREIMAGE_LENGTH),
+  publicOutput: Provable.Array(Field, 32),
   methods: {
-    Theta: {
+    ethereum: {
       privateInputs: [],
-      method(input: Field[][]) {
-        return theta(input);
+      method(preImage) {
+        return Keccak.ethereum(preImage);
       },
     },
-    PiRho: {
+    // No need for preNist Keccak_256 because it's identical to ethereum
+    nistSha3: {
       privateInputs: [],
-      method(input: Field[][]) {
-        return piRho(input);
-      },
-    },
-    Chi: {
-      privateInputs: [],
-      method(input: Field[][]) {
-        return chi(input);
-      },
-    },
-    Iota: {
-      privateInputs: [],
-      method(input: Field[][]) {
-        return iota(input, ROUND_CONSTANTS[0]);
-      },
-    },
-    Round: {
-      privateInputs: [],
-      method(input: Field[][]) {
-        return round(input, ROUND_CONSTANTS[0]);
-      },
-    },
-    BlockTransformation: {
-      privateInputs: [],
-      method(input: Field[][]) {
-        return blockTransformation(input);
+      method(preImage) {
+        return Keccak.nistSha3(256, preImage);
       },
     },
   },
 });
 
-// constraintSystem.fromZkProgram(
-//   KeccakBlockTransformation,
-//   'BlockTransformation',
-//   print
-// );
+await Keccak256.compile();
 
-console.log('KECCAK_TEST_STATE: ', KECCAK_TEST_STATE.toString());
+await equivalentAsync(
+  {
+    from: [array(uint(8), PREIMAGE_LENGTH)],
+    to: array(uint(8), 32),
+  },
+  { runs: RUNS }
+)(
+  (x) => {
+    let uint8Array = new Uint8Array(x.map(Number));
+    let result = keccak_256(uint8Array);
+    return Array.from(result).map(BigInt);
+  },
+  async (x) => {
+    let proof = await Keccak256.ethereum(x);
+    return proof.publicOutput;
+  }
+);
 
-console.log('Compiling...');
-await KeccakBlockTransformation.compile();
-console.log('Done!');
-console.log('Generating proof...');
-let proof0 = await KeccakBlockTransformation.BlockTransformation(
-  KECCAK_TEST_STATE
+await equivalentAsync(
+  {
+    from: [array(uint(8), PREIMAGE_LENGTH)],
+    to: array(uint(8), 32),
+  },
+  { runs: RUNS }
+)(
+  (x) => {
+    let thing = x.map(Number);
+    let result = sha3_256(new Uint8Array(thing));
```

Comment:
Oh so this is using another implementation of the Keccaks to compare against what the gadget provides? Nice

I wonder if it could make sense to create tests using some of the test vectors proposed by the Keccak Team. Maybe not to be run inside CI all the time, but at least having run them once before shipping it for production.